modules/internal/DefaultRectangular.chpl

```diff
@@ -842,24 +841,31 @@ module DefaultRectangular {
           sum += (ind(i) - off(i)) * blk(i) / abs(str(i)):idxType;
         return sum;
       } else {
-        var sum = if earlyShiftData then 0:idxType else origin;
-
-        // If the user asserts that there is no slicing in their program,
-        // then blk(rank) == 1. Knowing this, we need not multiply the final
-        // ind(...) by anything. This may lead to performance improvements for
-        // array accesses.
-        if assertNoSlicing {
-          for param i in 1..rank-1 {
-            sum += ind(i) * blk(i);
+        if (rank == 1 && earlyShiftData) {
+          if __primitive("optimize_array_blk_mult") {
+            return ind(1);
+          } else {
+            return ind(1) * blk(1);
```

Comment:
I'm definitely OK with leaving it in, particularly with the rationale noted here (though noting it in the comments would be even better.  

It does make me wonder whether there would be similar improvements we could do to the multi-dim case...  I assume all of this was with denormalization off?

Anyway, no need to do anything beyond a comment in the code for my sake.